mergin/report.py

```diff
@@ -0,0 +1,237 @@
+            self.operation = "insert"
+            self.old_geom = None
+            self.new_geom = changeset_entry.new_values[geom_idx]
+
+        self.count = None
+        self.length = None
+        self.area = None
+
+        if self.geom_type == "LINESTRING":
+            # we calculate change in length, for attributes changes only we set to 0
+            self.metric = "length"
+            if self.operation == "delete":
+                self.length = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.length = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.length = self.measure(self.new_geom)
+        elif self.geom_type == "POLYGON":
+            # we calculate change in area, for attributes changes only we set to 0
+            self.metric = "area"
+            if self.operation == "delete":
+                self.area = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.area = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.area = self.measure(self.new_geom)
+        else:
+            # regardless of geometry change count as 1
+            self.metric = "count"
+            self.count = 1
+
+    def measure(self, geom):
+        """ Return length or area of geometry based on type """
+        # calculate geom length/area only if QGIS API is available
+        try:
+            from qgis.core import QgsGeometry, QgsDistanceArea, QgsCoordinateReferenceSystem, QgsCoordinateTransformContext
+        except ImportError:
+            return -1
+
+        d = QgsDistanceArea()
+        d.setEllipsoid('WGS84')
+        crs = QgsCoordinateReferenceSystem()
+        crs.createFromString(self.crs)
+        d.setSourceCrs(crs, QgsCoordinateTransformContext())
+        g = QgsGeometry()
+        wkb_header_length = parse_gpkgb_header_size(geom)
+        wkb_geom = geom[wkb_header_length:]
+        g.fromWkb(wkb_geom)
+        if self.metric == "length":
+            return d.measureLength(g)
+        elif self.metric == "area":
+            return d.measureArea(g)
+        else:
+            return 1
+
+
+class ChangesetReport:
+    """ Report (aggregate) from geopackage changeset (diff file) """
+    def __init__(self, changeset_reader, schema):
+        self.cr = changeset_reader
+        self.schema = schema
+        self.entries = []
+        # let's iterate through reader and populate entries
+        for entry in self.cr:
+            schema_table = next((t for t in schema if t["table"] == entry.table.name), None)
+            # get geometry index in both gpkg schema and diffs values
+            geom_idx = next((index for (index, col) in enumerate(schema_table["columns"]) if col["type"] == "geometry"),
+                            None)
+            if not geom_idx:
+                continue
```

Comment:
does it mean that non-spatial tables are ignored?

and will this work in case geometry column has index zero? Shouldn't we use `if geom_ids is None` ?